### PR TITLE
Structural Preparations for PoA Support

### PR DESCRIPTION
**Pull request by holgerd77**
This PR provides some ground for a future PoA support, see #689. I am doing this now since I hope that this will allow us for a future PoA feature addition without the need for another major release respectively introducing breaking changes. I have done some reading along on the Clique PoA [EIP-225](https://eips.ethereum.org/EIPS/eip-225) and I think this should get a solid enough structural setup with this PR which should allow for non-breaking integration (by adding signer validation on PoA chains along an activated `Blockchain.validateConsensus` setting), this might benefit from some validation though.

The PR introduces code parts to now throw on `Block` or `Blockchain` validation triggers on setups which we just don't support yet (validate blocks on 'goerli') or which doesn't make sense (run 'validateDifficulty()' on goerli) and therefore reduces undefined behavior within the libraries.

@jochem-brouwer if this PR is supported and reviewed, we can merge on top of your blockchain refactoring changes with me rebasing or the the other way around, as you prefer.

### CODE REVIEW

**Comment by holgerd77** [comment]
Some tests are failing, so relabelling this as `WIP`, can be picked up though.

**Comment by holgerd77** [comment]
Ok, this was an easy fix 😄 , ready again for review.

File: packages/blockchain/src/index.ts
Comment (jochem-brouwer) [inline packages/blockchain/src/index.ts:100]: When we implement Clique we might want to figure out how we handle this here on a more abstract matter. Otherwise, we'd also create a `_clique?: Clique` object; it would make sense to pack those together in a more common interface, like a `Consensus` class?

File: packages/blockchain/src/index.ts
Comment (jochem-brouwer) [inline packages/blockchain/src/index.ts:160]: These checks are at multiple places, we might want to create a helper function for these tests?

File: packages/blockchain/src/index.ts
Comment (jochem-brouwer) [inline packages/blockchain/src/index.ts:80]: I am slightly confused here what the difference is between `validateConsensus` and `validateBlocks`. So if we use PoW and verify Consensus then it checks if the ethash of the block is correct? What does `validateBlocks` then do? Does it also imply that we verify the consensus, or can we validate the block but not the consensus? `validateBlocks` would throw on things like some parameters which are incorrect (like a timestamp which is less than the parent block)?

File: packages/blockchain/src/index.ts
Comment (rumkin) [inline packages/blockchain/src/index.ts:396]: According to constructor condition there couldn't be `_validateConsensus` without `_ethash` instance, so the condition could be simplified to just check `_validateConsensus` to be true.

**Comment by jochem-brouwer** [comment]
Added some questions and some nitpicks. But looks great in general 😄

File: packages/blockchain/src/index.ts
Comment (rumkin) [inline packages/blockchain/src/index.ts:153]: Maybe it would be reasonable to allow users to provide ethash instance with options? It's using in examples to let user optimizae blockchain work, so it seems rational to accept users instance, rather than let get into the guts?

**Review by rumkin — CHANGES_REQUESTED** [review]

File: packages/blockchain/src/index.ts
Comment (holgerd77) [reply packages/blockchain/src/index.ts:100]: Yeah, I didn't find a good abstraction for it now and found it overblown to dig further this this is just a preparatory PR for some time in the future. On the first round I now added the `_` to more clearly indicate that this is a member which shouldn't be used from the outside. Think a structure should become more clear once we start with e.g. Clique integration.

File: packages/blockchain/src/index.ts
Comment (holgerd77) [reply packages/blockchain/src/index.ts:160]: I wanted to have slightly different error messages depending on where the errors occur, that's the reasoning for not having this unified.

File: packages/blockchain/src/index.ts
Comment (holgerd77) [reply packages/blockchain/src/index.ts:139]: Have updated (next push). Have to say that I really have to learn this left-hand-side right-hand-side thing here (did it the wrong way around on first try) 😀 . This is (a bit) the somewhat more dangerous part of this operator, but I guess this won't be a problem any more and easily spotted once we are all more a bit used to it.

File: packages/blockchain/src/index.ts
Comment (holgerd77) [reply packages/blockchain/src/index.ts:146]: Yes, that's a good point. We have an open issue for that at #702, this has (somewhat) high demand and this might still be a candidate to integrate along the dev releases. I wouldn't want to side-introduce along the work here, is it ok if we do this as some separate work along the issue from above? I will directly go over, update the issue and place some open questions I still have on this.

Comment (holgerd77) [reply packages/blockchain/src/index.ts:146]: See https://github.com/ethereumjs/ethereumjs-vm/issues/702#issuecomment-722296424

File: packages/blockchain/src/index.ts
Comment (holgerd77) [reply packages/blockchain/src/index.ts:153]: I think this is something for a dedicated issue. I would put this under the kinds of feature requests though which we should only tackle if there is a demand expressed, otherwise we run into the risk of implementing things just on guess which at the end are not really used.

File: packages/blockchain/src/index.ts
Comment (holgerd77) [reply packages/blockchain/src/index.ts:396]: Technically you are right, I just decided to keep this for now to limit a bit the risk that this check is forgotten once the library is expanded with other consensus types. I will put this check separately into the next line to make this more clear on next push.

File: packages/common/src/index.ts
Comment (holgerd77) [reply packages/common/src/index.ts:640]: I just took over the suboptimal structure from the other methods from for e.g. `hardforks()` and the like 😛 , if we improve here I think we should do in a separate PR and then for all functions affected.

I think we shouldn't spent to much time on this now - there are more pressing things - but if there is an easy way that would be great, @rumkin eventually you can take this on at some point after merge here? I am still not such a `TypeScript` pro on some things, have e.g. not yet worked with generics in some meaningful way and it should be easier for you to do I guess.

Comment (holgerd77) [reply packages/common/src/index.ts:640]: (or if you don't want to do in the next days or so maybe just open a small issue here)

File: packages/vm/examples/run-blockchain/index.ts
Comment (holgerd77) [reply packages/vm/examples/run-blockchain/index.ts:24]: Yeah, that's a good point, have added.

File: packages/blockchain/src/index.ts
Comment (holgerd77) [reply packages/blockchain/src/index.ts:80]: Yes, I have to say that I am somewhat confused by this as well and I have to say that I don't have a good overview which kind of these validation checks are absolutely necessary and which ones are just a nice-to-have thing. I think we'll get some more intuitive understanding on this over time when dealing more with integrating the `blockchain`, `block` and `tx` libraries in the client. Actually love that this work triggers theses kind of questions.

I think one crucial basic part is the underdeveloped documentation here. On this round I will concentrate on this and go through the validation methods and give this some more clarity. Not sure yet, it might even be worth to rename the one or the other method, some good names can sometimes really do miracles on these kind of things.

File: packages/block/src/header.ts
Comment (holgerd77) [inline packages/block/src/header.ts:390]: I think this was not a good choice to completely throw on block validation and prevent users to be able to do all the remaining checks. I will change this to just skip the difficulty check.

File: packages/blockchain/src/index.ts
Comment (holgerd77) [reply packages/blockchain/src/index.ts:80]: This was actually a super-good question which triggered direct changes in the code, thanks! 😄

**Comment by holgerd77** [comment]
Ok, this is now ready for re-review, for an overview on the changes pushed see 60ecd32fbdd818923e14e85c6d573f4531148edf, thanks @jochem-brouwer and @rumkin for all the valuable suggestions.

I took a lot of time on giving the validation methods a better documentation, I think this should make this a lot clearer on first-time read, I also learned a lot myself here. 😄  I also made the `validate` methods for `Block` and `Blockchain` less strict again and don't completely throw on non-PoW chains but just skip the difficulty check and do everything else, think this makes more sense.

File: packages/blockchain/src/index.ts
Comment (jochem-brouwer) [reply packages/blockchain/src/index.ts:80]: I think in general, this is the idea: `validatePow` was used to validate the mining work, while `validateBlock` is to essentially validate everything else in the block (the header is correct, the uncles are correct, parent hash exists in the chain, etc.). I assume this is because sometimes you just want to execute a block and skip the PoW check since it is very expensive to come up with a correct PoW. Also in terms of the test suite. `ethereum/tests` also has many tests which have `NoProof`, which implies that you check everything except PoW.

File: packages/block/src/header.ts
Comment (jochem-brouwer) [reply packages/block/src/header.ts]: This is OK for this PR, but I think this is what we should do:

We introduce this new `BlockValidator` interface, which does two things: it has `validateConsensus(block: Block)` which allows us to validate consensus of this `Block` (for Ethash, this is to verify the PoW of this block. For Clique, it is to do the POA check). We also add `validateBlock(block: Block)`, which does these checks like validate difficulty, timestamp, etc. For instance, in Clique, it seems that the `difficulty` is used as an extra parameter which doesn't have much to do with difficulty in general. [See the Clique EIP 225](https://eips.ethereum.org/EIPS/eip-225). Note: we probably also want to add ways to input a `BlockHeader` here besides `Block`.

We might want to either figure out what these Consensus methods have in common (probably; timestamp is always strictly larger). But most things are actually Consensus specific: for instance, Clique doesn't have any uncle validation logic, besides that it should be a constant (empty list hash it seems). 

So in general the suggestion is that we remove all PoW consensus/validation logic from `Blockchain` and `Block` and instead encapsulate this in an abstract class `Consensus`, where we can add both Ethash/Clique and possibly more algorithms in the future. So this would be a strategy pattern here, it would make a lot of sense to me.

Looks like these changes are rather important but also breaking, so we should discuss these asap if we want to implement them before V5.

File: packages/blockchain/src/index.ts
Comment (jochem-brouwer) [reply packages/blockchain/src/index.ts:98]: Small remark here; if the VM only uses it in the tests then I think it is OK to just cast the `blockchain` object as any in order to get `_ethash`.

**Comment by jochem-brouwer** [comment]
OK this is great! I also love that you added a lot of documentation here.

I think the way to proceed (see [this comment](https://github.com/ethereumjs/ethereumjs-vm/pull/937#discussion_r518187942)) is to (in a new PR) remove most of this consensus logic from blockchain/block and use this strategy pattern so we can choose which strategy we use for validating the block/block header. So these concrete instances would be a PoW consensus and a Clique consensus. These would allow us to validate the consensus (PoW/Clique) and also validate the block/block header (note that in Clique there are no uncles, for instance, so the uncle hash is always the empty list hash).

**Comment by jochem-brouwer** [comment]
@holgerd77 I suggest we merge this one after my block/blockchain PRs are merged so you can rebase on top of my changes (seems to be less work than the other way around).

File: packages/block/src/header.ts
Comment (holgerd77) [reply packages/block/src/header.ts]: I wouldn't want to take this in any more, this is getting too large and too risky for a short-term last minute refactor. This PR wasn't meant to open up this whole "how to implement Clique?" topic but just doing some base preparations. Let's please see how far we'll get with this along the v5 release series, but I rather would want to close this whole PR than going here any further.

**Comment by holgerd77** [comment]
Currently blocked by #942

**Comment by holgerd77** [comment]
Will take the next hour or so to update (hopefully rebase) with the changes from #942 and previous block/blockchain touching PRs.

**Comment by holgerd77** [comment]
Bildschirmfoto 2020-11-10 um 09 27 25

Oh, that's a tough one. 😂

**Comment by holgerd77** [comment]
Ok, this is now ready for re-review. 😄  //cc @jochem-brouwer

**Review by jochem-brouwer — APPROVED** [review]
LGTM! 😄

**Pull request merged**
